### ISSUE TITLE
Generic type parameter is lost after generic member call

### ISSUE THREAD

**Issue by alexfoxgill** [microsoft/TypeScript#11584]
This is the simplest version of this bug I could find:

``` TypeScript
function example<T1>() {
    let x = {
        foo: <T2>(t2: T2) => x,
        bar: (t1: T1) => x
    }
    return x;
}

example<number>()
    .foo("hello")
    .bar(1) // Argument of type 'number' is not assignable to parameter of type 'T1'
```

The call to `bar` fails to compile. The type is `example<number>`, so it should compile.

If I try to add type information to `x`, another compilation error occurs:

``` TypeScript
type Example<T1> = {
    foo: <T2>(t2: T2) => Example<T1>,
    bar: (t1: T1) => Example<T1>
}
function example<T1>() {
    let x: Example<T1> = {
        foo: <T2>(t2: T2): Example<T1> => x,
        bar: (t1: T1): Example<T1> => x
    }
    return x;
}

example<number>()
    .foo("hello")
    .bar(1)
```

> Type '{ foo: <T2>(t2: T2) => Example<T1>; bar: (t1: T1) => Example<T1>; }' is not assignable to type 'Example<T1>'.
>   Types of property 'foo' are incompatible.
>     Type '<T2>(t2: T2) => Example<T1>' is not assignable to type '<T2>(t2: T2) => Example<T1>'.
>       Type 'Example' is not assignable to type 'Example<T1>'.
>         Types of property 'bar' are incompatible.
>           Type '(t1: T1) => Example' is not assignable to type '(t1: T1) => Example<T1>'.
>             Types of parameters 't1' and 't1' are incompatible.
>               Type 'T1' is not assignable to type 'T1'.
> let x: Example<T1>

Interestingly, removing either the `foo` or `bar` methods will allow the code to successfully compile:

``` TypeScript
type Example<T1> = {
    foo: <T2>(t2: T2) => Example<T1>,
    //bar: (t1: T1) => Example<T1>
}
function example<T1>() {
    let x: Example<T1> = {
        foo: <T2>(t2: T2): Example<T1> => x,
        //bar: (t1: T1): Example<T1> => x
    }
    return x;
}
```

``` TypeScript
type Example<T1> = {
    //foo: <T2>(t2: T2) => Example<T1>,
    bar: (t1: T1) => Example<T1>
}
function example<T1>() {
    let x: Example<T1> = {
        //foo: <T2>(t2: T2): Example<T1> => x,
        bar: (t1: T1): Example<T1> => x
    }
    return x;
}
```

**Comment by alexfoxgill** [comment]
Classes do not have this issue:

```TypeScript
class Example<TFoo> {
    constructor(foo: TFoo) {
    }

    foo(f: TFoo) {
        return this;
    }

    bar<TBar>(b: TBar) {
        return this;
    }
}

new Example(0)
    .bar("hello")
    .foo(1)
```

**Comment by Andarist** [comment]
This has already been fixed: [TS playground](https://www.typescriptlang.org/play?jsx=0&ts=5.3.0-dev.20230825&filetype=ts#code/GYVwdgxgLglg9mABAUwB4EMC2AHANsgHgBUBGAPgAoBKRAbwChEnF8pFVEBeOx5v4OHABciYgCZKUMSKJianMuwA0vPkwBG6AE4iKUEjJLzFqVYgC+ZrciggtSVAG56l+mix5CYEJnXItlFRmAHQCcBQARAAWyLi4cBFBfMGaWhRGiEA), cc @jakebailey

**Comment by jakebailey** [comment]
The error changed in #10676 for 2.0 to instead mention `1`, then fully fixed in #13356 for 2.1. Though, I'm not totally sure we have a test for this, given what changed in the latter PR.
